### PR TITLE
Rewrite BoundsInitializer

### LINKED ISSUE #6373: Performance regression after 3.34.0
I'm seeing some invocations of the nullness checker take significantly longer on versions of the checker framework after 3.34.0.

I minimized an example below. The invocation time more than doubles between 3.34.0 and 3.35.0. The performance for newer versions seems roughly consistent with 3.35.0.

```
$ time ./checker-framework-3.34.0/checker/bin/javac -processor nullness X.java
real    0m42.137s
user    1m9.228s
sys     0m2.893s

$ time ./checker-framework-3.35.0/checker/bin/javac -processor nullness X.java
real    1m56.321s
user    2m25.035s
sys     0m3.416s
```

```java
public class X {

  abstract static class C1<
          C extends C1<C, Q, B, D, CR>,
          Q extends C2<C, Q, B, D, CR>,
          B extends C3<C, Q, B, D, CR>,
          D extends C4<C, Q, B, D, CR>,
          CR extends C5<CR>>
      extends C6 {}

  static class C6 {}

  abstract static class C2<
          C extends C1<C, Q, B, D, RT>,
          Q extends C2<C, Q, B, D, RT>,
          B extends C3<C, Q, B, D, RT>,
          D extends C4<C, Q, B, D, RT>,
          RT extends C5<RT>>
      implements C7 {}

  abstract static class C3<
      C extends C1<C, Q, B, D, R>,
      Q extends C2<C, Q, B, D, R>,
      B extends C3<C, Q, B, D, R>,
      D extends C4<C, Q, B, D, R>,
      R extends C5<R>> {}

  abstract static class C4<
      C extends C1<C, Q, B, D, R>,
      Q extends C2<C, Q, B, D, R>,
      B extends C3<C, Q, B, D, R>,
      D extends C4<C, Q, B, D, R>,
      R extends C5<R>> {
    interface I<T> {}
  }

  abstract static class C5<R2 extends C5<R2>> implements C7 {}

  interface C7 {}

  abstract static class C8<
      C extends C1<C, Q, B, D, CR>,
      Q extends C2<C, Q, B, D, CR>,
      B extends C3<C, Q, B, D, CR>,
      D extends C4<C, Q, B, D, CR>,
      CR extends C5<CR>,
      RpT extends C5<RpT>> {

    public static <
            C extends C1<C, Q, B, D, CR>,
            Q extends C2<C, Q, B, D, CR>,
            B extends C3<C, Q, B, D, CR>,
            D extends C4<C, Q, B, D, CR>,
            CR extends C5<CR>,
            RpT extends C5<RpT>>
        Builder<C, Q, B, D, CR, RpT> n(Q q) {
      throw new AssertionError();
    }

    abstract static class Builder<
        C extends C1<C, Q, B, D, CR>,
        Q extends C2<C, Q, B, D, CR>,
        B extends C3<C, Q, B, D, CR>,
        D extends C4<C, Q, B, D, CR>,
        CR extends C5<CR>,
        RpT extends C5<RpT>> {

      public abstract Builder<C, Q, B, D, CR, RpT> f(C9<?> x);

      public C8<C, Q, B, D, CR, RpT> b() {
        throw new AssertionError();
      }
    }
  }

  abstract static class C9<W extends C9<W>> {}

  static final class C10 extends C9<C10> {}

  abstract static class C11<W extends C9<W>, B extends C11<W, B>> {}

  static final class C12 extends C11<C10, C12> {

    public C10 b() {
      throw new AssertionError();
    }
  }

  static class C13 {

    public static final C12 n() {
      return new C12();
    }

    static final class C14 extends C1<C14, C15, C16, C17, C18> {}

    static final class C15 extends C2<C14, C15, C16, C17, C18> {}

    static final class C18 extends C5<C18> {}

    static class C17 extends C4<C14, C15, C16, C17, C18> implements C4.I<Long>, C19 {}

    static final class C16 extends C3<C14, C15, C16, C17, C18> {

      public C15 b() {
        throw new AssertionError();
      }
    }

    static final C16 q() {
      throw new AssertionError();
    }
  }

  interface C19 {}

  void f() {
    var x = C8.n(C13.q().b()).f(C13.n().b()).b();
  }
}
```

### PR DESCRIPTION
**Pull request by smillst**
Fixes #6373

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/type/BoundsInitializer.java [hunk 1]
```diff
@@ -1,1388 +1,298 @@
...
-            ((AnnotatedWildcardType) atmArg).setTypeArgOfRawType();
-          }
-        }
-      } else {
-        typeArgs = declaredType.typeArgs;
+    private void initializeTypeVariable(AnnotatedTypeVariable annotatedTypeVariable) {
+      TypeVariable t = annotatedTypeVariable.getUnderlyingType();
+      if (!annotatedTypeVariable.isDeclaration()) {
+        t = (TypeVariable) TypeAnnotationUtils.unannotatedType(t);
+        typeVariableMap.put(t, annotatedTypeVariable);
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/type/BoundsInitializer.java:154]: Is method `initializeTypeVariable` only ever called once?  If not, then this seems to be redoing work and overwriting a value in the map.  I have the same question for the method below, `initializeWildcard`.

**Review by mernst — CHANGES_REQUESTED** [review]
I have some questions that can be clarified in documentation.
Thanks!

File: framework/src/main/java/org/checkerframework/framework/type/BoundsInitializer.java [hunk 2]
```diff
@@ -1,1388 +1,298 @@
...
- * initializer method is called for a particular bound, the entirety of that bound, including
- * circular references, will be created.
+ * variables and wildcards. It ensures that recursive type variables refer to themselves at the
+ * correct location.
+ *
+ * <p>Its static helper methods are called from AnnotatedTypeMirror. When an initializer method is
+ * called for a particular bound, the entirety of that bound, including circular references, is
+ * created.
  */
 public class BoundsInitializer {
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/type/BoundsInitializer.java:44]: I wonder whether "BoundsInitializer" is the right name for this.  It copies the structure of a type (more specifically, a type variable or wildcard) into an AnnotatedTypeMirror.  Could the name reflect its behavior?  "Initializer" is not very evocative.

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeMirror.java [hunk 3]
```diff
@@ -1086,18 +1086,46 @@ public void setTypeArguments(List<? extends AnnotatedTypeMirror> ts) {
...
+        }
       } else {
-        // Initialize type argument for a non-raw declared type that has type arguments/
-        BoundsInitializer.initializeTypeArgs(this);
-        return typeArgs;
+        for (TypeMirror javaTypeArg : t.getTypeArguments()) {
+          AnnotatedTypeMirror typeArg =
+              AnnotatedTypeMirror.createType(javaTypeArg, atypeFactory, false);
+          typeArgs.add(typeArg);
+        }
```
[the remark this answers is not in the document]
Comment (smillst) [reply framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeMirror.java:1126]: None of the methods in `AnnotatedTypeMirror` add annotations to their component types.

**Review by mernst — DISMISSED** [review]

**Review by mernst — APPROVED** [review]
Thanks.

**Pull request merged**